stability_image/image.go

```diff
@@ -0,0 +1,392 @@
+package stability_image
+
+import (
+	"bytes"
+	"github.com/disintegration/imaging"
+	"github.com/esimov/stackblur-go"
+	"github.com/foobaz/lossypng/lossypng"
+	"github.com/lucasb-eyer/go-colorful"
+	"github.com/mazznoer/colorgrad"
+	"golang.org/x/image/webp"
+	"image"
+	"image/color"
+	"image/gif"
+	"image/jpeg"
+	"image/png"
+	"log"
+	"math/rand"
+	"os"
+	"strconv"
+)
+
+var (
+	MaxDimension     uint64 = 1536
+	MinDimension     uint64 = 256
+	MaxPixels        uint64 = 960 * 1024
+	MinPixels        uint64 = 768 * 768
+	DefaultDimension uint64 = 768
+	DimensionStep    uint64 = 64
+)
+
+func EncodePng(
+	i image.Image,
+	level png.CompressionLevel,
+) (
+	encoded *[]byte,
+	err error,
+) {
+	encoded = &[]byte{}
+	buf := new(bytes.Buffer)
+
+	encoder := png.Encoder{CompressionLevel: level}
+	writeErr := encoder.Encode(buf, i)
+	if writeErr != nil {
+		return nil, err
+	}
+	*encoded = buf.Bytes()
+	return encoded, nil
+}
+
+func DecodeImage(raw *[]byte) (
+	i image.Image,
+	format string,
+	dim *image.Point,
+	err error) {
+	i, format, err = image.Decode(bytes.NewReader(*raw))
+	if err != nil {
+		return nil, "", dim, err
+	}
+	dimVal := i.Bounds().Size()
+	return i, format, &dimVal, err
+}
+
+func Fill(img *image.RGBA, col *color.RGBA) {
+	bounds := img.Bounds()
+	for x := bounds.Min.X; x < bounds.Max.X; x++ {
+		for y := bounds.Min.Y; y < bounds.Max.Y; y++ {
+			img.Set(x, y, col)
```

Comment:
Instead of calling a function (Set) for each pixel which has to do multiply with row stride each time can you get pointer to the start of row and index from there? This is how it's commonly done in C/C++ for example.
```
for (int y = 0; y < height; y++) {
    color* row = image.ptr<color>(y);
    for (int x = 0; x < width; x++) {
        row[x] = ...
    }
}
```
Then writes go sequentially through memory using cache nicely and no extra function calls or multiplications per pixel.

Reply:
Maybe, but not without using `unsafe`, I think. Let me think about it.